python/lsst/daf/butler/registries/remote.py

```diff
@@ -0,0 +1,522 @@
+
+        # Storage class information should be pulled from server.
+        # Dimensions should be pulled from server.
+
+    def __str__(self) -> str:
+        return str(self._db)
+
+    def __repr__(self) -> str:
+        return f"RemoteRegistry({self._db!r}, {self.dimensions!r})"
+
+    def isWriteable(self) -> bool:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        # Can be used to prevent any PUTs to server
+        return self._writeable
+
+    def copy(self, defaults: Optional[RegistryDefaults] = None) -> Registry:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        if defaults is None:
+            # No need to copy, because `RegistryDefaults` is immutable; we
+            # effectively copy on write.
+            defaults = self.defaults
+        return type(self)(self._db, defaults, self.isWriteable())
+
+    @property
+    def dimensions(self) -> DimensionUniverse:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        if self._dimensions is not None:
+            return self._dimensions
+
+        # Access /dimensions.json on server and cache it locally.
+        response = httpx.get(str(self._db.join("universe")))
+        response.raise_for_status()
+
+        config = DimensionConfig.fromString(response.text, format="json")
+        self._dimensions = DimensionUniverse(config)
+        return self._dimensions
+
+    def refresh(self) -> None:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+
+        # Need to determine what to refresh.
+        # Might need a server method to return all the DatasetTypes up front.
+        pass
+
+    @contextlib.contextmanager
+    def transaction(self, *, savepoint: bool = False) -> Iterator[None]:
+        # Transaction handling for client server is hard and will require
+        # some support in the server to store registry changes and defer
+        # committing them. This will likely require a change in transaction
+        # interface. For now raise.
+        raise NotImplementedError()
+
+    # insertOpaqueData + fetchOpaqueData + deleteOpaqueData
+    #    There are no managers for opaque data in client. This implies
+    #    that the server would have to have specific implementations for
+    #    use by Datastore. DatastoreBridgeManager also is not needed.
+
+    def registerCollection(self, name: str, type: CollectionType = CollectionType.TAGGED,
+                           doc: Optional[str] = None) -> None:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        raise NotImplementedError()
+
+    @functools.lru_cache
+    def getCollectionType(self, name: str) -> CollectionType:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        # This could use a local cache since collection types won't
+        # change.
+        path = f"v1/registry/collection/type/{name}"
+        response = httpx.get(str(self._db.join(path)))
```

Comment:
Does `ButlerURI` enforce that the server URI ends in a `/`? If not, you may want to enforce that.

Reply:
It's annoyingly complicated in general but yes, ButlerURI knows if it represents a "directory".